hachoir/parser/image/cr2.py

```diff
@@ -0,0 +1,110 @@
+"""
+Canon CR2 raw image data, version 2.0 image parser.
+
+Authors: Fernando Crespo
+Creation date: 21 february 2017
+"""
+
+from hachoir.parser import Parser
+from hachoir.field import SeekableFieldSet, RootSeekableFieldSet, Bytes, String, UInt8, UInt16, UInt32
+from hachoir.core.endian import LITTLE_ENDIAN, BIG_ENDIAN
+from hachoir.core.text_handler import textHandler, hexadecimal
+from hachoir.parser.image.exif import ParserError, IFD, IFD_TAGS
+
+
+def getStrips(ifd):
+    data = {}
+    for i, entry in enumerate(ifd.array('entry')):
+        data[entry['tag'].display] = entry
+    # image data
+    if "StripOffsets" in data and "StripByteCounts" in data:
+        offs = ifd.getEntryValues(data["StripOffsets"])
+        bytes = ifd.getEntryValues(data["StripByteCounts"])
+        for off, byte in zip(offs, bytes):
+            yield off.value, byte.value
+
+
+class ImageFile(SeekableFieldSet):
+
+    def __init__(self, parent, name, description, ifd):
+        SeekableFieldSet.__init__(self, parent, name, description, None)
+        self._ifd = ifd
+
+    def createFields(self):
+        for off, byte in getStrips(self._ifd):
+            self.seekByte(off, relative=False)
+            yield Bytes(self, "strip[]", byte)
+
+
+class CR2File(RootSeekableFieldSet, Parser):
+    PARSER_TAGS = {
+        "id": "cr2",
+        "category": "image",
+        "file_ext": ("cr2",),
+        "mime": ("image/x-canon-cr2",),
+        "min_size": 15,
+        "magic": ((b"CR", 8),),
+        "description": "Canon CR2 raw image data, version 2.0"
+    }
+
+    # Correct endian is set in constructor
+    endian = LITTLE_ENDIAN
+
+    def __init__(self, stream, **args):
+        RootSeekableFieldSet.__init__(
+            self, None, "root", stream, None, stream.askSize(self))
+        if self.stream.readBytes(0, 2) == b"MM":
+            self.endian = BIG_ENDIAN
+        Parser.__init__(self, stream, **args)
+
+    def validate(self):
+        endian = self.stream.readBytes(0, 2)
+        if endian not in (b"MM", b"II"):
+            return "Invalid endian (%r)" % endian
+        if self["version"].value != 42:
+            return "Unknown Canon TIFF version - " + str(self["version"].value)
+        if self["cr_identifier"].value != "CR":
+            return "Unknown Canon Raw File"
+        return True
+
+    def createFields(self):
+        iff_start = self.absolute_address + self.current_size
+        yield String(self, "endian", 2, "Endian ('II' or 'MM')", charset="ASCII")
+        if self["endian"].value not in ("II", "MM"):
+            raise ParserError("Invalid endian!")
```

Comment:
I used the `tiff.py` as a template and the class `TIFF` in `exif.py` has this check. I'll remove that code.